torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,134 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// TODO(osalpekar): Tune RetryDuration to be something in seconds per
+// team recommendations. Kept at 100ms for the sake of Milestone 1.
+const auto kRpcRetryDuration = std::chrono::milliseconds(100);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+const float kBackoff = 1.2;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(true);
+  }
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(false);
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+}
 
-RpcAgent::~RpcAgent() = default;
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+  auto future = std::make_shared<FutureMessage>();
+  auto fm = send(to, std::move(message));
+  fm->addCallback([future](
+                      const rpc::Message& message,
+                      const c10::optional<utils::FutureError>& futErr) {
+    if (futErr) {
+      LOG(INFO) << "Send try 1 failed\n";
```

Comment:
We need to differentiate failure conditions which are retry-able, vs. not.

A network connection dropped is very different from an unrecoverable function/level-level error

This might not be completely trivial given how unstructured some of our errors are, but I expect that we want to have a function somewhere to tell us whether a given error is retryable. I suspect we want a whitelist (e.g. connection dropped)

Regardless, there are definitely app-level errors that we want to propagate here, and the retry mechanism needs to interpret ones it doesn't understand as such.

Reply:
If we only call `sendWithRetries` on internal messages will this ensure that we're not retrying needlessly? My concern with a whitelist is that we'll have to add all possible network/IO related exceptions and we might miss some and constantly have to update.